lib/Tools/GenericSwizzling.cpp

```diff
@@ -0,0 +1,327 @@
+      segment.push_back(1 << b); // free variables
+  if (segment.size() >= lenSegment) {
+    segment.resize(lenSegment);
+    return segment;
+  }
+
+  // A and B are the difference sets
+  SmallVector<int32_t> A, B;
+  for (int32_t v : setSrc)
+    if (!setDst.contains(v))
+      A.push_back(v);
+  for (int32_t v : setDst)
+    if (!setSrc.contains(v))
+      B.push_back(v);
+  if (A.size() > B.size()) {
+    std::swap(A, B);
+  }
+  llvm::sort(A);
+  llvm::sort(B);
+  // A is the smaller set now
+  auto logBankConflicts = std::min<int32_t>(
+      std::max<int32_t>(0, lenSegment - A.size() - segment.size()), A.size());
+  // Conflict-free
+  for (int i = logBankConflicts; i < A.size(); ++i)
+    segment.push_back(A[i] ^ B[i]);
+  // Write conflicts
+  segment.append(A.begin(), A.begin() + logBankConflicts);
+  // Read conflicts
+  segment.append(B.begin(), B.begin() + logBankConflicts);
+
+  if (segment.size() > lenSegment)
+    segment.resize(lenSegment);
+  return segment;
+}
+
+SmallVector<int32_t> complementBasis(ArrayRef<int32_t> basis, int32_t rank) {
+  const int32_t nRows = basis.size();
+  // Build an nRows × rank matrix: each row is one of your 'basis' vectors.
+  auto mat = std::make_unique<uint64_t[]>(nRows);
+  for (int r = 0; r < nRows; ++r)
+    mat[r] = static_cast<uint64_t>(basis[r]);
+
+  // RREF that, with 'rows = nRows' and 'cols = rank'.
+  f2reduce::inplace_rref_strided(mat.get(), /*rows=*/nRows,
+                                 /*cols=*/rank, /*stride=*/1);
+
+  // Collect which coordinate-columns have pivots.
+  llvm::SmallDenseSet<int32_t> pivotCols;
+  for (int r = 0; r < nRows; ++r) {
+    if (mat[r]) {
+      pivotCols.insert(__builtin_ctzll(mat[r])); // leading-1 position
+    }
+  }
+
+  // Now any coordinate i ∈ [0,rank) that *isn’t* in pivotCols is free,
+  // so the unit-vector (1<<i) belongs in your complement.
+  SmallVector<int32_t> comp;
+  for (int i = 0; i < rank; ++i)
+    if (!pivotCols.contains(i))
+      comp.push_back(1 << i);
+
+  return comp;
+}
+} // anonymous namespace
+
+namespace mlir::triton::gpu {
+
+SmallVector<int32_t> intersectionBasis(ArrayRef<int32_t> b1,
+                                       ArrayRef<int32_t> b2, int32_t rank) {
+  auto ns1 = nullspaceBasis(b1, rank);
+  auto ns2 = nullspaceBasis(b2, rank);
+  auto joint = llvm::to_vector(llvm::concat<int32_t>(ns1, ns2));
+  auto result = nullspaceBasis(joint, rank);
+  return result;
+}
+
+LinearLayout optimalSwizzling(const LinearLayout &src, const LinearLayout &dst,
+                              int32_t bitwidth) {
+  assert(llvm::equal(src.getInDimNames(), dst.getInDimNames()) &&
+         "src and dst must have identical in dims");
+  assert(llvm::equal(src.getOutDims(), dst.getOutDims()) &&
+         "src and dst must have identical out dims and shape");
+  assert((bitwidth > 0) && llvm::isPowerOf2_32(bitwidth) &&
+         "bitwidth must be a power of two");
+
+  const int32_t rank = src.getTotalOutDimSizeLog2();
```

Comment:
outSize or size might be more intuitive

Reply:
changed it to dim as it is the dimension of the vector space we are working with.